internal/scheduler/gang_scheduler.go

```diff
@@ -0,0 +1,121 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	schedulerconstraints "github.com/armadaproject/armada/internal/scheduler/constraints"
+	schedulercontext "github.com/armadaproject/armada/internal/scheduler/context"
+	"github.com/armadaproject/armada/internal/scheduler/nodedb"
+	"github.com/armadaproject/armada/internal/scheduler/schedulerobjects"
+)
+
+// GangScheduler schedules one gang at a time. GangScheduler is not aware of queues.
+type GangScheduler struct {
+	constraints       schedulerconstraints.SchedulingConstraints
+	schedulingContext *schedulercontext.SchedulingContext
+	nodeDb            *nodedb.NodeDb
+}
+
+func NewGangScheduler(
+	sctx *schedulercontext.SchedulingContext,
+	constraints schedulerconstraints.SchedulingConstraints,
+	nodeDb *nodedb.NodeDb,
+) (*GangScheduler, error) {
+	return &GangScheduler{
+		constraints:       constraints,
+		schedulingContext: sctx,
+		nodeDb:            nodeDb,
+	}, nil
+}
+
+func (sch *GangScheduler) Schedule(ctx context.Context, gctx *schedulercontext.GangSchedulingContext) (ok bool, unschedulableReason string, err error) {
+	defer func() {
+		// If any job in a gang fails to schedule, set the unschedulableReason for all jobs in the gang,
+		// and remove it from the scheduling context.
+		if unschedulableReason == "" {
```

Comment:
surely the check here should be against `ok` not `unschedulableReason` else the jobs won't be evicted if `ok` is false but `unschedulableReason` is the empty string?

Reply:
More generally it's not immediately clear to me what comination of values is ok here.  What does it mean if err is non nill but ok is true?!  Generally this isn't a concern for callers because err trumps all, but here you've got a deferred function that is taking action without considering err.  It seems dangerous to mutate state in the case that some unknown error may or may not have occurred?

Reply:
I've changed it to be
```
	defer func() {
		// If any job in a gang fails to schedule, set the unschedulableReason for all jobs in the gang,
		// and remove it from the scheduling context.
		if err == nil && !ok {
			for _, jctx := range gctx.JobSchedulingContexts {
				sch.schedulingContext.EvictJob(jctx.Job)
				jctx.UnschedulableReason = unschedulableReason
			}
			sch.schedulingContext.AddGangSchedulingContext(gctx)
		}
	}()
```